### DISCUSSION TITLE
Memory management question and odd behavior with larger files

### DISCUSSION THREAD

**Discussion by GM-Script-Writer-62850** [micropython/micropython#9550] [General]
Using a PICO W with rp2-pico-w-20220923-unstable-v1.19.1-443-g30e50ab19.uf2

I noticed some strange behavior when opening/closing files and running `gc.collect` after and checking `gc.mem_free`
lets say i open a file of say 3457 bytes and dump it over http and look at my ram then clean it up and look again, the space look to have been freed up that i just used, however if i open say 70294 bytes and assuming i do not run out of ram the ram does not get freed as I expected based on the test with the smaller file

i am/was opening files like this
```
from gc import mem_free as fram, collect as clearRAM
...
# Using a asynchronous web server, see page 23: https://datasheets.raspberrypi.com/picow/connecting-to-the-internet-with-pico-w.pdf
...
			size=stat(file)[6]
			print('Open',file,'(',size,') with', fram())
			if(size*1.1 < fram()): # end*1.1 is to be sure we are not completely out of memory, in theory you do not need the *1.1
				writer.write(header+
					'\r\nContent-Language: pt-BR\r\nCache-Control: max-age='+str(rng(604800,1209600))+'\r\n'+
					'Content-Length:'+str(size)+
					'\r\n\r\n'
				)
				with open(file, 'rb') as f:
					writer.write(f.read())
				print("File closed; free ram:",fram())
				clearRAM()
				print("Garbage collected; free ram:",fram())
			else:
				# This is gonna be a issue if it happens to index.html
				print("Low Ram: Offload",file,"to alt server")
				if compress != False:
					file=compress# revert file to the non gzip version
				writer.write('HTTP/1.0 301 OK\r\nLocation: http://10.0.0.25:8080/thermostat/'+file[4:])
```
Sometimes i have have ~100K free and run out of ram opening a ~70K file: `MemoryError: memory allocation failed, allocating 41472 bytes` how am i only getting ~41K when asking for ~70K when i have ~100K

**Comment by karfas** [comment]
This behavior is caused by memory fragmentation (see https://forum.micropython.org/viewtopic.php?t=1260 for a discussion).
For your problem at hand, you could read smaller chunks of the file, e.g. 
```
with open(file, 'rb') as f:
    b = f.read(size = 1024)
    while(b):
         writer.write(b)
         b = f.read(size = 1024)
```

**Reply by GM-Script-Writer-62850** [reply]
basically what i have now

```
with open(file, 'rb') as f:
	while True:
		chunk=f.read(1024)
		if not len(chunk):
			break
		writer.write(chunk)
```

that said i like `while(b)` better, so 
```
size=1024
with open(file, 'rb') as f:
	chunk = f.read(size)
		while(chunk):
			writer.write(chunk)
			chunk = f.read(size)
```

**Reply by GM-Script-Writer-62850** [reply]
interesting, so basically if you were to try to allocate ram based on free ram space you would need to de-fragment the ram first never would have thought ram would care about that with it being solid state, now i know why it was not working as expected

**Reply by peterhinch** [reply]
A random observation: this is the classic use case for the walrus operator:
```python
size=1024
with open(file, 'rb') as f:
    while(chunk := f.read(size)):
        writer.write(chunk)
```

**Reply by GM-Script-Writer-62850** [reply]
This still eats ram and does not release/overwrite it, and i have to clear ram inside the while loop

**Reply by jimmo** [reply]
Peter is right about the walrus operator making the code simpler. It's unrelated to the memory allocation.

The problem is that read() has to allocate a buffer for each chunk.
If you want to avoid the allocation, use readinto with a reusable buffer.

Here's an example of doing exactly this:
https://github.com/micropython/micropython-lib/blob/master/micropython/mip/mip.py#L33 (This code could also use the walrus operator :) )

**Reply by GM-Script-Writer-62850** [reply]
still not perfect, took longer to get it to happen
```
				print("free RAM:",fram())
				with open(file, 'rb') as f:
					print("free RAM:",fram())
					buf = memoryview(bytearray(1024))
					print("free RAM:",fram())
					while(c := f.readinto(buf)):
						writer.write(buf)
```

```
Open www/style/images/snowing.gif ( 70294 ) with 109792
free RAM: 108928
free RAM: 107728
free RAM: 106672
Task exception wasn't retrieved
future: <Task> coro= <generator object 'webUI' at 20014890>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "<stdin>", line 490, in webUI
  File "uasyncio/stream.py", line 1, in write
MemoryError: memory allocation failed, allocating 18620 bytes
```

**Reply by jimmo** [reply]
You need to drain the writer, other wise it might buffer inside Stream.

See https://github.com/micropython/micropython/blob/master/extmod/uasyncio/stream.py#L66

Also your code needs to do what the code i linked to does with the memory view and the length returned from readinto, otherwise the last chunk will be too long.

**Reply by GM-Script-Writer-62850** [reply]
`await write.drain()` is done later in the script when the connection is closed or do i need to do it in the loop?
```
				print("free RAM:",fram())
				with open(file, 'rb') as f:
					print("free RAM:",fram())
					buf = memoryview(bytearray(s))
					print("free RAM:",fram())
					while(c := f.readinto(buf)):
						writer.write(buf if c == s else buf[:c])
.....
	try:
		await writer.drain()
		await writer.wait_closed()
		print("Client disconnected")
	except Exception as e:
		print("Error:",e)
```
`writer.write(buf if c == s else buf[:c])` is that what you meant? cause it still fails

i do not really understand what i am doing
i think i am allocating 1024 bytes of ram, dumping 1024 bytes into it, writing it, overwriting that memory with the next 1024 bytes then wiring that and so on till i run out of data

i do not know why it is trying to allocate over 10x my buffer size and running out of ram

**Reply by peterhinch** [reply]
You need to issue `.drain` immediately after `.write` otherwise, as @jimmo says, it will create an unreasonably large buffer. Unfortunately `uasyncio` does not have an unbuffered `.write`: I wrote a [PR](https://github.com/micropython/micropython/pull/7868) to address this but it has not been accepted. So some allocation still occurs, but it is much reduced.

Please see [this tutorial reference](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#63-using-the-stream-mechanism) for guidance on stream I/O.

**Reply by GM-Script-Writer-62850** [reply]
Thanks, BTW is there a version of `urequests` that uses `uasyncio`? i just need http get/post request for exporting data to my server, what i have is fine as it with `urequests` but it could be marginally better

**Reply by jimmo** [reply]
https://github.com/micropython/micropython-lib/tree/master/micropython/uaiohttpclient

**Reply by GM-Script-Writer-62850** [reply]
Thanks, but that does not look to have post data support, maybe i can patch urequests...
[urequest_untested_uasyncio_patch.zip](https://github.com/micropython/micropython/files/9757399/urequest_untested_uasyncio_patch.zip)
maybe it would be a bad idea to do this with a connection limit, the pico w can only hold 4 at a time right? since my use case is only on my local lan and the reply is probably just going to be 2 chars long

**Reply by GM-Script-Writer-62850** [reply]
I made a version of urequests that uses uasyncio; https://github.com/micropython/micropython-lib/issues/550 maybe it could be done better IDK, i am no python expert
